agent/eni/networkutils/utils_linux.go

```diff
@@ -17,22 +17,113 @@
+	macAddressBackoffMultiple = 1.5
+)
+
+// macAddressRetriever is used to retrieve the mac address of a device. It collects
+// all the information necessary to start this operation and stores the result in
+// the 'macAddress' attribute
+type macAddressRetriever struct {
+	dev           string
+	netlinkClient netlinkwrapper.NetLink
+	macAddress    string
+	// timeout specifies the timeout duration before giving up when
+	// looking for an ENI's mac address on the host
+	timeout time.Duration
+	ctx     context.Context
+}
+
 // GetMACAddress retrieves the MAC address of a device using netlink
-func GetMACAddress(dev string, netlinkClient netlinkwrapper.NetLink) (string, error) {
-	dev = filepath.Base(dev)
-	link, err := netlinkClient.LinkByName(dev)
+func GetMACAddress(ctx context.Context,
+	timeout time.Duration,
+	dev string,
+	netlinkClient netlinkwrapper.NetLink) (string, error) {
+	retriever := &macAddressRetriever{
+		dev:           dev,
+		netlinkClient: netlinkClient,
+		ctx:           ctx,
+		timeout:       timeout,
+	}
+	return retriever.retrieve()
+}
+
+// retrieve retrives the mac address of a network device. If the retrieved mac
+// address is empty, it retries the operation with a timeout specified by the
+// caller
+func (retriever *macAddressRetriever) retrieve() (string, error) {
+	backoff := utils.NewSimpleBackoff(macAddressBackoffMin, macAddressBackoffMax,
+		macAddressBackoffJitter, macAddressBackoffMultiple)
+	ctx, cancel := context.WithTimeout(retriever.ctx, retriever.timeout)
+	defer cancel()
+
+	err := utils.RetryWithBackoffCtx(ctx, backoff, func() error {
+		retErr := retriever.retrieveOnce()
+		if retErr != nil {
+			seelog.Warnf("Unable to retrieve mac address for device '%s': %v",
+				retriever.dev, retErr)
+			return retErr
+		}
+
+		if retriever.macAddress == "" {
+			seelog.Debugf("Empty mac address for device '%s'", retriever.dev)
+			// Return a retriable error when mac address is empty. If the error
+			// is not wrapped with the RetriableError interface, RetryWithBackoffCtx
+			// treats them as retriable by default
+			return errors.Errorf("eni mac address: retrieved empty address for device %s",
+				retriever.dev)
+		}
+
+		return nil
+	})
 	if err != nil {
 		return "", err
 	}
-	return link.Attrs().HardwareAddr.String(), err
+	// RetryWithBackoffCtx returns nil when the context is cancelled. Check if there was
+	// a timeout here. TODO: Fix RetryWithBackoffCtx to return ctx.Err() on context Done()
+	if err = ctx.Err(); err != nil {
+		return "", errors.Wrapf(err, "eni mac address: timed out waiting for eni device '%s'",
+			retriever.dev)
+	}
+
+	return retriever.macAddress, nil
+}
+
+// retrieveOnce retrieves the MAC address of a device using netlink.LinkByName
+func (retriever *macAddressRetriever) retrieveOnce() error {
+	dev := filepath.Base(retriever.dev)
+	link, err := retriever.netlinkClient.LinkByName(dev)
+	if err != nil {
+		return utils.NewRetriableError(utils.NewRetriable(false), err)
```

Comment:
Are you sure that we should never retry in this case?

Reply:
yeah, an error means a syscall failed for describing the device. at this point of time, it doesn't make sense to retry it as it really should be visible on the instance